Circuit for opcode `BALANCE`

silathdiir opened the pull request:
Spec: https://github.com/privacy-scaling-explorations/zkevm-specs/pull/248

### Summary

1. Update `BALANCE` bus-mapping and add circuit.
2. Fix to lookup `Balance` if account exists, otherwise lookup `NonExisting`.
3. Add a `HashSet` to save non-existent account addresses in both bus-mapping and circuit `ExeStep`s.

~~Not confirm if necessary to constrain "If the given account doesn't exist, then it will push 0 onto the stack instead." in the circuit.
And is there a method (could be used in circuit) to lookup if account is existing? Thanks.~~

~~Blocked by PR https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/907~~

CPerezz commented:
Blocked by https://github.com/privacy-scaling-explorations/zkevm-specs/pull/248 which is at the same time blocked by https://github.com/privacy-scaling-explorations/zkevm-specs/issues/249

silathdiir commented:
Will try to update this PR since https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/907 has been merged.

silathdiir commented:
Hi @CPerezz , I updated this `BALANCE` circuit with non-existing proofs. Could you help review again when you have time? Thanks.

CPerezz commented on bus-mapping/src/circuit_input_builder/execution.rs:
```diff
@@ -44,6 +45,8 @@ pub struct ExecStep {
     pub bus_mapping_instance: Vec<OperationRef>,
     /// Error generated by this step
     pub error: Option<ExecError>,
+    /// Non existent account addresses
+    pub non_existent_accounts: HashSet<Address>,
```
For now this looks fine. But IMO we will be able to mix both existant and non-existant accounts within a single HashSet and structure.

CPerezz commented on zkevm-circuits/src/evm_circuit/execution/balance.rs:
```diff
@@ -0,0 +1,302 @@
+            .assign(region, offset, Value::known(F::from(is_warm)))?;
+
+        let exists = step.account_exists(&address);
+        let balance = if exists {
+            block.rws[step.rw_indices[5]]
+                .table_assignment_aux(block.randomness)
+                .value
```
I'm unsure why do we need `block.randomness` to set the balance. Is it passed directly as an RLC??

silathdiir replied:
I checked function [table_assignment_aux](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/witness/rw.rs#L334). But it seems that only `value` is needed here, so I replaced it with `value_assignment(block.randomness)`.

And reference function [value_assignment](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/witness/rw.rs#L518), it seems that the argument `randomness` is used in function `random_linear_combine` as below:
```
AccountFieldTag::CodeHash | AccountFieldTag::Balance => {
    RandomLinearCombination::random_linear_combine(value.to_le_bytes(), randomness)
}
```

silathdiir replied:
Hi @CPerezz , could you help review this again when you have time? Thanks.

CPerezz replied:
Ye sorry @silathdiir I missed this!!!

CPerezz reviewed (commented):
Just wondering the balance passed in the assignation of the Opcode circuit impl.

For the rest this looks good. If you cold provide feedback, I'll approve immediately!

CPerezz reviewed (approved):
LGTM! I'd want another review from someone at Scroll before merging this! @icemelon or @silathdiir could you ping someone? :smile:

silathdiir commented:
> LGTM! I'd want another review from someone at Scroll before merging this! @icemelon or @silathdiir could you ping someone? 😄

Yes. I will ping others to help review. Thanks a lot.

DreamWuGit commented on zkevm-circuits/src/evm_circuit/execution/balance.rs:
```diff
@@ -0,0 +1,300 @@
+            + (1.expr() - is_warm.expr()) * GasCost::COLD_ACCOUNT_ACCESS.expr();
+
+        let step_state_transition = StepStateTransition {
+            rw_counter: Delta(cb.rw_counter_offset()),
```
` rw_counter: Delta(cb.rw_counter_offset()),` not good style since we don't know exactly rw counter change, suggest to use normal way  ` rw_counter: 7.expr()`

DreamWuGit commented on bus-mapping/src/circuit_input_builder/execution.rs:
```diff
@@ -44,6 +45,8 @@ pub struct ExecStep {
     /// Error generated by this step
     pub error: Option<ExecError>,
+    /// Non existent account addresses
+    pub non_existent_accounts: HashSet<Address>,
```
most probably do not need this, for circuit side, existing can be check by reading buss mapping record,  for `AccountFieldTag::NonExisting`, no need additional step info.

silathdiir replied:
Yes exactly 🙏, I fixed to get both `balance` and `exists` in circuit function `assign_exec_step` as below:
```
let (balance, exists) = match block.rws[step.rw_indices[5]] {
    Rw::Account {
        field_tag: AccountFieldTag::Balance,
        value,
        ..
    } => (
        RandomLinearCombination::random_linear_combine(
            value.to_le_bytes(),
            block.randomness,
        ),
        true,
    ),
    Rw::Account {
        field_tag: AccountFieldTag::NonExisting,
        ..
    } => (F::zero(), false),
    _ => unreachable!(),
};
```

DreamWuGit commented on zkevm-circuits/src/witness/step.rs:
```diff
@@ -40,6 +34,8 @@ pub struct ExecStep {
     /// The opcode corresponds to the step
     pub opcode: Option<OpcodeId>,
+    /// Non existent account addresses
+    pub non_existent_accounts: HashSet<Address>,
```
same as above comment

silathdiir replied:
Deleted.

DreamWuGit commented on zkevm-circuits/src/evm_circuit/execution/balance.rs:
```diff
@@ -0,0 +1,300 @@
+
+        let address = block.rws[step.rw_indices[0]].stack_value().to_address();
+        let mut address_bytes = address.0;
+        address_bytes.reverse();
```
want to know what reason for `reverse` it

silathdiir replied:
Try to investigate `ethers_core::types::Address` and code in `eth-types`, it seems that the `Address` is big endian.

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/eth-types/src/lib.rs#L167

lispc commented on zkevm-circuits/src/evm_circuit/execution/balance.rs:
```diff
@@ -0,0 +1,313 @@
+
+        let exists = cb.query_bool();
+        let balance = cb.condition(exists.expr(), |cb| {
+            let balance = cb.query_cell();
```
can the "query_cell" be moved out of the closure? It is some "structure/allocation" info, better not be guarded with "if" branch.

silathdiir replied:
Moved out `query_cell` as:
```
let balance = cb.query_cell();
let exists = cb.query_bool();
cb.condition(exists.expr(), |cb| {
    cb.account_read(
        from_bytes::expr(&address.cells),
        AccountFieldTag::Balance,
        balance.expr(),
    );
});
```

lispc commented on zkevm-circuits/src/evm_circuit/execution/balance.rs:
```diff
@@ -0,0 +1,313 @@
+        cb.stack_push(select::expr(exists.expr(), balance.expr(), 0.expr()));
+
+        let gas_cost = is_warm.expr() * GasCost::WARM_ACCESS.expr()
+            + (1.expr() - is_warm.expr()) * GasCost::COLD_ACCOUNT_ACCESS.expr();
```
use select::expr?

silathdiir replied:
Fixed to:
```
let gas_cost = select::expr(
    is_warm.expr(),
    GasCost::WARM_ACCESS.expr(),
    GasCost::COLD_ACCOUNT_ACCESS.expr(),
);
```

lispc commented:
need to modify https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3da1888cfdea1f05062477279d9c5088e01fac19/zkevm-circuits/src/witness/step.rs#L184

silathdiir commented:
> need to modify
> 
> https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3da1888cfdea1f05062477279d9c5088e01fac19/zkevm-circuits/src/witness/step.rs#L184

Sorry for the wrong revert. Fixed to `OpcodeId::BALANCE => ExecutionState::BALANCE` and moved to next one of `ADDRESS`.
